src/Resources/contao/widgets/PickerWidget.php

```diff
@@ -0,0 +1,240 @@
+ *
+ * @property boolean        $multiple
+ * @property array          $rootNodes
+ * @property string         $fieldType
+ * @property string         $foreignTable
+ * @property string         $context
+ * @property string         $labelField
+ * @property string         $icon
+ * @property boolean        $sortable
+ * @property array|callable $label_callback
+ *
+ * @author Leo Feyer <https://github.com/leofeyer>
+ * @author Kamil Kuzminski <https://github.com/qzminski>
+ */
+class PickerWidget extends \Widget
+{
+
+	/**
+	 * Submit user input
+	 * @var boolean
+	 */
+	protected $blnSubmitInput = true;
+
+	/**
+	 * Template
+	 * @var string
+	 */
+	protected $strTemplate = 'be_widget';
+
+
+	/**
+	 * Load the database object
+	 *
+	 * @param array $arrAttributes
+	 */
+	public function __construct($arrAttributes=null)
+	{
+		$this->import('Database');
+		parent::__construct($arrAttributes);
+	}
+
+
+	/**
+	 * Return an array if the "multiple" attribute is set
+	 *
+	 * @param mixed $varInput
+	 *
+	 * @return mixed
+	 */
+	protected function validator($varInput)
+	{
+		$this->checkValue($varInput);
+
+		if ($this->hasErrors())
+		{
+			return '';
+		}
+
+		// Return the value as usual
+		if ($varInput == '')
+		{
+			if ($this->mandatory)
+			{
+				$this->addError(sprintf($GLOBALS['TL_LANG']['ERR']['mandatory'], $this->strLabel));
+			}
+
+			return '';
+		}
+		elseif (strpos($varInput, ',') === false)
+		{
+			return $this->multiple ? array(intval($varInput)) : intval($varInput);
+		}
+		else
+		{
+			$arrValue = array_map('intval', array_filter(explode(',', $varInput)));
+
+			return $this->multiple ? $arrValue : $arrValue[0];
+		}
+	}
+
+
+	/**
+	 * Check the selected value
+	 *
+	 * @param mixed $varInput
+	 */
+	protected function checkValue($varInput)
+	{
+		if ($varInput == '' || !is_array($this->rootNodes) || !$this->Database->fieldExists('pid', $this->foreignTable))
+		{
+			return;
+		}
+
+		if (strpos($varInput, ',') === false)
+		{
+			$arrIds = array(intval($varInput));
+		}
+		else
+		{
+			$arrIds = array_map('intval', array_filter(explode(',', $varInput)));
+		}
+
+		if (count(array_diff($arrIds, array_merge($this->rootNodes, $this->Database->getChildRecords($this->rootNodes, $this->foreignTable)))) > 0)
+		{
+			$this->addError($GLOBALS['TL_LANG']['ERR']['invalidRecords']);
+		}
+	}
+
+
+	/**
+	 * Generate the record label
+	 *
+	 * @param array $data
+	 *
+	 * @return string
+	 */
+	protected function generateRecordLabel(array $data)
+	{
+		// Label callback
+		if (is_array($this->label_callback))
+		{
+			$callback = $this->label_callback;
+
+			$label = static::importStatic($callback[0])->{$callback[1]}($data, $this->objDca);
+		}
+		elseif (is_callable($callback = $this->label_callback))
+		{
+			$label = $callback($data, $this->objDca);
+		}
+		elseif ($this->labelField)
+		{
+			// Label field
+			$label = $data[$this->labelField];
+		}
+		else
+		{
+			$label = $data['id'];
+		}
+
+		// Add the icon
+		if ($this->icon)
```

Comment:
Don't you think it would make sense to put the icon in the callback as well? E.g. the page picker has different icons per page type and this sounds like a typical requirement to me :)

Reply:
This `icon` setting is defined by the developer and there is no icon provided by default. Thus if your icon is not static you can just not set this property and add it in the `label_callback`.

Also the `label_callback` is called before the evaluation of the `icon` setting, which allows you to use callback just to format the label properly but still set the static icon afterwards.

Reply:
I see, I can live with that :)